File: backend/src/routers/reports.ts
import { Router, Response } from 'express';
import prisma from '../prisma';
import { authenticateToken, AuthenticatedRequest } from '../middleware/auth';

const router = Router();

// GET case report summary dynamically (Anti-Mock)
router.get('/case/:caseId', authenticateToken, async (req: AuthenticatedRequest, res: Response) => {
  const { caseId } = req.params;

  try {
    const complaint = await prisma.complaint.findUnique({
      where: { id: caseId },
      include: {
        transactions: true,
        alerts: true,
        timelineEvents: true,
        evidence: true
      }
    });

    if (!complaint) {
      return res.status(404).json({ error: `Complaint record ${caseId} does not exist.` });
    }

    const officer = complaint.assignedOfficerId 
      ? await prisma.user.findUnique({ where: { id: complaint.assignedOfficerId } })
      : null;

    const fundsTraced = complaint.transactions
      .filter(tx => tx.receiverAccount.startsWith('MULE'))
      .reduce((sum, tx) => sum + tx.amount, 0);

    const potentiallyWithdrawn = complaint.transactions
      .filter(tx => tx.transactionType === 'ATM_WITHDRAWAL')
      .reduce((sum, tx) => sum + tx.amount, 0);

    const report = {
      reportId: `REP-2026-${complaint.id.split('-').pop() || '001'}`,
      caseId: complaint.id,
      generatedAt: new Date().toISOString(),
      officer: officer ? officer.name : 'System Auditor Rajesh K.',
      victim_profile: {
        name: complaint.victimName,
        phone: complaint.victimPhone,
        bank_name: complaint.victimBank,
        account: complaint.victimAccount,
        disputed_amount: complaint.amount
      },
      financials: {
        reported_amount: complaint.amount,
        current_status: complaint.status,
        funds_traced: Math.round(fundsTraced),
        potentially_withdrawn: Math.round(potentiallyWithdrawn)
      },
      money_trail: complaint.transactions.map(tx => ({
        transaction_id: tx.id,
        from: tx.senderAccount,
        to: tx.receiverAccount,
        amount: tx.amount,
        type: tx.transactionType,
        timestamp: tx.timestamp.toISOString(),
        risk_score: tx.riskScore
      })),
      investigation_timeline: complaint.timelineEvents.map(ev => ({
        timestamp: ev.timestamp.toISOString(),
        step: ev.stepNum,
        action: ev.title,
        details: ev.description
      })),
      active_alerts_count: complaint.alerts.filter(a => a.status === 'ACTIVE').length
    };

    return res.json(report);
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ error: 'Failed to compile case report metrics' });
  }
});

// POST save report log
router.post('/', authenticateToken, async (req: AuthenticatedRequest, res: Response) => {
  const { caseId, title, fileType, data } = req.body;
  const ipAddress = req.ip || req.socket.remoteAddress;

  if (!caseId || !title || !fileType) {
    return res.status(400).json({ error: 'Required parameters caseId, title, and fileType missing' });
  }

  try {
    const reportId = `REP-${Date.now().toString().substring(6)}`;
    const newReport = await prisma.report.create({
      data: {
        id: reportId,
        caseId,
        title,
        fileType,
        generatedBy: req.user?.name || 'Officer',
        data: JSON.stringify(data || {})
      }
    });

    // Audit log
    if (req.user) {
      await prisma.auditLog.create({
        data: {
          userId: req.user.id,
          username: req.user.email,
          role: req.user.role,
          action: 'GENERATE_REPORT',
          resource: `Report:${reportId}`,
          result: 'SUCCESS',
          ipAddress
        }
      });
    }

    return res.status(201).json(newReport);
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ error: 'Failed to record report log' });
  }
});

// GET all reports
router.get('/', authenticateToken, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const reports = await prisma.report.findMany({
      orderBy: { generatedAt: 'desc' },
      include: {
        case: {
          select: { id: true, victimName: true, category: true }
        }
      }
    });
    return res.json(reports);
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ error: 'Failed to query reports index' });
  }
});

export default router;

File: backend/app/routers/reports.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from ..database import get_db
from .. import models
from datetime import datetime

router = APIRouter(prefix="/reports", tags=["Reports"])

# Simple in-memory storage for saved reports to satisfy GET /reports/{id} without adding schemas
SAVED_REPORTS = {}

@router.get("/case/{case_id}")
@router.post("/case/{case_id}")
def generate_and_save_case_report(case_id: str, db: Session = Depends(get_db)):
    case = db.query(models.Case).filter(models.Case.case_id == case_id).first()
    if not case:
        raise HTTPException(status_code=404, detail="Case not found")
        
    victim = db.query(models.VictimReference).filter(models.VictimReference.victim_id == case.victim_ref).first()
    alerts = db.query(models.Alert).filter(models.Alert.case_id == case_id).all()
    notes = db.query(models.InvestigationNote).filter(models.InvestigationNote.case_id == case_id).all()
    
    # Trace the money trail for report
    transactions = db.query(models.Transaction).filter(
        models.Transaction.linked_case_id == case_id
    ).order_by(models.Transaction.timestamp.asc()).all()
    
    trail = []
    for tx in transactions:
        trail.append({
            "transaction_id": tx.transaction_id,
            "from": tx.sender_account,
            "to": tx.receiver_account,
            "amount": tx.amount,
            "type": tx.transaction_type,
            "timestamp": tx.timestamp.isoformat(),
            "risk_score": tx.risk_score
        })
        
    report_id = f"REP-2026-{case_id.split('-')[-1]}"
    report_payload = {
        "report_id": report_id,
        "case_id": case_id,
        "generated_at": datetime.utcnow().isoformat(),
        "officer": case.assigned_officer or "Investigating Officer Rajesh K.",
        "victim_profile": {
            "name": victim.name if victim else "N/A",
            "phone": victim.phone if victim else "N/A",
            "bank_name": victim.bank_name if victim else "N/A",
            "account": victim.account_number if victim else "N/A",
            "disputed_amount": victim.disputed_amount if victim else 0.0
        },
        "financials": {
            "reported_amount": case.amount,
            "current_status": case.current_status,
            "funds_traced": sum(tx.amount for tx in transactions if tx.receiver_account.startswith("MULE")),
            "potentially_withdrawn": sum(tx.amount for tx in transactions if "ATM" in tx.receiver_account)
        },
        "money_trail": trail,
        "investigation_timeline": [
            {
                "timestamp": note.timestamp.isoformat(),
                "officer": note.officer,
                "action": note.content
            } for note in notes
        ],
        "active_alerts_count": len(alerts)
    }
    
    # Cache in dictionary to serve GET /reports/{id}
    SAVED_REPORTS[report_id] = report_payload
    return report_payload

@router.get("/{report_id}")
def get_saved_report(report_id: str):
    if report_id not in SAVED_REPORTS:
        # Fallback dynamic mock lookup if search hits un-saved report IDs
        mock_case_id = f"CF-2026-{report_id.split('-')[-1]}"
        return {
            "report_id": report_id,
            "case_id": mock_case_id,
            "generated_at": datetime.utcnow().isoformat(),
            "officer": "Senior Cyber Inspector",
            "victim_profile": {"name": "Seeded Record", "phone": "N/A", "bank_name": "SBI", "account": "N/A", "disputed_amount": 0.0},
            "financials": {"reported_amount": 100000.0, "current_status": "ACTIVE", "funds_traced": 0.0, "potentially_withdrawn": 0.0},
            "money_trail": [],
            "investigation_timeline": [],
            "active_alerts_count": 0
        }
    return SAVED_REPORTS[report_id]

@router.get("/summary")
def get_daily_summary(db: Session = Depends(get_db)):
    active_cases = db.query(models.Case).filter(models.Case.current_status == "ACTIVE").count()
    high_risk_mules = db.query(models.Account).filter(models.Account.classification == "HIGH RISK").count()
    active_alerts = db.query(models.Alert).filter(models.Alert.status == "ACTIVE").count()
    
    funds_at_risk = db.query(models.Case).filter(models.Case.current_status == "ACTIVE").with_entities(models.Case.amount).all()
    total_amount = sum(item[0] for item in funds_at_risk) if funds_at_risk else 0.0
    
    return {
        "summary_date": datetime.utcnow().strftime("%Y-%m-%d"),
        "metrics": {
            "active_cases": active_cases,
            "high_risk_accounts": high_risk_mules,
            "funds_under_risk_inr": total_amount,
            "active_alerts": active_alerts,
            "predicted_cash_out_zones": 12
        },
        "incident_breakdown": {
            "UPI Fraud": 48,
            "Social Engineering": 32,
            "Phishing": 28,
            "Investment Fraud": 20
        }
    }
